File: articles/networks-systems-biological-complexity/r/module_summary.R
# Module-level network summary.

edge_path <- file.path("data", "biological_network_edges.csv")
node_path <- file.path("data", "biological_network_nodes.csv")

if (!file.exists(edge_path)) {
  edge_path <- file.path("..", "data", "biological_network_edges.csv")
  node_path <- file.path("..", "data", "biological_network_nodes.csv")
}

edges <- read.csv(edge_path)
nodes <- read.csv(node_path)

module_lookup <- setNames(nodes$module, nodes$node)

edges$source_module <- module_lookup[edges$source]
edges$target_module <- module_lookup[edges$target]
edges$within_module <- edges$source_module == edges$target_module

overall <- data.frame(
  total_edges = nrow(edges),
  within_module_edges = sum(edges$within_module),
  between_module_edges = sum(!edges$within_module),
  within_module_fraction = mean(edges$within_module)
)

module_summary <- aggregate(
  weight ~ source_module + target_module,
  data = edges,
  FUN = mean
)

edge_counts <- aggregate(
  source ~ source_module + target_module,
  data = edges,
  FUN = length
)

names(edge_counts)[names(edge_counts) == "source"] <- "n_edges"

module_summary <- merge(module_summary, edge_counts)

print(round(overall, 5))
print(module_summary[order(-module_summary$n_edges), ])

File: articles/networks-systems-biological-complexity/python/module_summary.py
"""
Module-level network summary.

Run:
    python python/module_summary.py
"""

from pathlib import Path

import pandas as pd

from network_complexity_core import module_summary


ARTICLE_DIR = Path(__file__).resolve().parents[1]
EDGE_PATH = ARTICLE_DIR / "data" / "biological_network_edges.csv"
NODE_PATH = ARTICLE_DIR / "data" / "biological_network_nodes.csv"


def main() -> None:
    edges = pd.read_csv(EDGE_PATH)
    nodes = pd.read_csv(NODE_PATH)

    node_modules = dict(zip(nodes["node"], nodes["module"]))
    summary = module_summary(edges, node_modules)

    overall = pd.DataFrame(
        {
            "total_edges": [len(edges)],
            "within_module_edges": [
                sum(node_modules[s] == node_modules[t] for s, t in zip(edges["source"], edges["target"]))
            ],
        }
    )
    overall["between_module_edges"] = overall["total_edges"] - overall["within_module_edges"]
    overall["within_module_fraction"] = overall["within_module_edges"] / overall["total_edges"]

    print(overall.round(5).to_string(index=False))
    print(summary.round(5).to_string(index=False))


if __name__ == "__main__":
    main()
